hooks/pre-push

```diff
@@ -10,30 +33,95 @@ GREEN='\033[0;32m'
 YELLOW='\033[0;33m'
 NC='\033[0m' # No Color
 
+PREFIX="${GREEN}[PREPUSH]${NC}"
+FAILURE="${RED}FAILED${NC}"
+WARNING="${YELLOW}[WARNING]${NC}"
+SKIPPED="${YELLOW}SKIPPED${NC}"
+SUCCESS="${GREEN}ok${NC}"
+
+printf "${PREFIX} Clean working copy ... "
 git diff --exit-code &>/dev/null
-if [ "$?" != 0 ];
-then
-	echo ${RED}ERROR${NC} You have uncommitted changes please commit or stash them before pushing so that I can run tests!
-	exit 1
+if [ "$?" == 0 ]; then
+	printf "${SUCCESS}\n"
+else
+	if [ "${TARPC_ALLOW_DIRTY}" == "1" ]
+	then
+		printf "${SKIPPED}\n"
+	else
+		printf "${FAILURE}\n"
+		exit 1
+	fi
 fi
 
-printf "${YELLOW}[PRESUBMIT]${NC} Running tests ... "
+PREPUSH_RESULT=0
+
+# args:
+# 1 - cargo command to run (build/test)
+# 2 - directory name of crate to build
+# 3 - rust toolchain (nightly/stable/beta)
+run_cargo() {
+	if [ "$1" == "build" ]; then
+		VERB=Building
+	else
+		VERB=Testing
+	fi
+	if [ "$3" != "" ]; then
+		printf "${PREFIX} $VERB $2 on $3 ... "
+		multirust run $3 cargo $1 --manifest-path $2/Cargo.toml &>/dev/null
+	else
+		printf "${PREFIX} $VERB $2 ... "
+		cargo $1 --manifest-path $2/Cargo.toml &>/dev/null
+	fi
+	if [ "$?" != "0" ]; then
+		printf "${FAILURE}\n"
+		PREPUSH_RESULT=1
+	else
+		printf "${SUCCESS}\n"
+	fi
+}
+
+TOOLCHAIN_RESULT=0
+check_toolchain() {
+	printf "${PREFIX} Checking for $1 toolchain ... "
+	if [[ $(multirust list-toolchain) =~ $1 ]]; then
+		printf "${SUCCESS}\n"
+	else
+		TOOLCHAIN_RESULT=1
+		PREPUSH_RESULT=1
+		printf "${FAILURE}\n"
+	fi
+}
+
+printf "${PREFIX} Checking for multirust ... "
+command -v multirust &>/dev/null
+if [ "$?" == 0 ] && [ "${TARPC_USE_CURRENT_TOOLCHAIN}" == "" ]; then
+	printf "${SUCCESS}\n"
+
+	check_toolchain stable
+	check_toolchain beta
+	check_toolchain nightly
+	if [ ${TOOLCHAIN_RESULT} == 1 ]; then
+		exit 1
+	fi
 
-TEST_RESULT=0
-cargo test --manifest-path tarpc/Cargo.toml &>/dev/null
-if [ "$?" != "0" ];
-then
-	printf "${RED}FAILED${NC}"
-	TEST_RESULT=1
+	run_cargo build tarpc stable
+	run_cargo build tarpc_examples stable
+
+	run_cargo build tarpc beta
+	run_cargo build tarpc_examples beta
+
+	run_cargo build tarpc nightly
+	run_cargo build tarpc_examples nightly
+
+	# We still rely on some nightly stuff for tests
+	run_cargo test tarpc nightly
+	run_cargo test tarpc_examples nightly
 else
-	printf "${GREEN}ok${NC}"
-fi
-printf "\n"
+	printf "${FAILURE}\n"
```

Comment:
"ok" is good for success, but maybe just not "FAILURE" when a toolchain is missing